File: R/trex_selector_methods/support_generators.R
# ==============================================================================
# support_generators.R
# ==============================================================================
#
# Support-index construction helpers for TRex simulation demos.
# Sourced by each demo file — do not run standalone.
#
# Contents:
#   make_support_capped_spread()    — Deterministic evenly-spaced support.
#   make_support_random()           — Uniformly random support.
#   make_support_bt_one_per_block() — One active predictor per BT block.
# ==============================================================================


# ==============================================================================
# make_support_capped_spread
# ==============================================================================

#' Build an evenly-spaced support with a capped gap (deterministic).
#'
#' Mirrors the C++ `capped_spread_support()`:
#'   gap = min(p %/% s, max_gap)
#'   indices = {1, 1 + gap, 1 + 2*gap, ..., 1 + (s-1)*gap}  (1-based)
#'
#' @param s       Number of active predictors.
#' @param p       Total number of predictors.
#' @param max_gap Maximum gap between consecutive active indices. Default 10.
#'
#' @return Integer vector of length s (1-based R indices).
#'
#' @examples
#' make_support_capped_spread(10, 1000, 20)
#' # -> 1 + c(0,20,40,...,180)  =  {1, 21, 41, ..., 181}
make_support_capped_spread <- function(s, p, max_gap = 10L) {
  gap <- min(p %/% s, max_gap)
  1L + (0L:(s - 1L)) * gap
}


# ==============================================================================
# make_support_random
# ==============================================================================

#' Build a uniformly random support without replacement (stochastic).
#'
#' Mirrors the C++ `random_support()`:
#'   set.seed(seed + 500000L)
#'   sample(p, s) sorted ascending (1-based)
#'
#' @param s    Number of active predictors.
#' @param p    Total number of predictors.
#' @param seed Integer seed.
#'
#' @return Integer vector of length s (1-based R indices), sorted ascending.
#'
#' @examples
#' make_support_random(5, 500, seed = 2026)
make_support_random <- function(s, p, seed) {
  set.seed(seed + 500000L)
  sort(sample.int(p, size = s, replace = FALSE))
}


# ==============================================================================
# make_support_bt_one_per_block
# ==============================================================================

#' Build the natural BT support: exactly one active predictor per block.
#'
#' Blocks are defined the same way as in `dgp_bt_snr()`: block m (1-based)
#' covers 1-based column indices [(m-1)*block_size + 1, m*block_size].
#'
#' @param p        Total number of predictors. Must be divisible by n_blocks.
#' @param n_blocks Number of blocks (equals the number of active predictors).
#' @param seed     If NULL (default), picks the FIRST predictor of each block
#'                 (deterministic).  If an integer, picks a random predictor
#'                 from each block using set.seed(seed).
#'
#' @return Sorted integer vector of length n_blocks with 1-based indices.
make_support_bt_one_per_block <- function(p, n_blocks, seed = NULL) {
  stopifnot(p %% n_blocks == 0L)
  block_size   <- p %/% n_blocks
  block_starts <- (seq_len(n_blocks) - 1L) * block_size  # 0-based block starts

  if (is.null(seed)) {
    # Deterministic: first predictor of each block
    offsets <- rep(0L, n_blocks)
  } else {
    # Random: one predictor drawn uniformly from each block
    set.seed(seed)
    offsets <- sample.int(block_size, n_blocks, replace = TRUE) - 1L
  }

  sort(block_starts + offsets + 1L)  # convert to 1-based
}

File: Python/trex_selector_methods/trex/support_generators.py
# ==============================================================================
# support_generators.py
# ==============================================================================
#
# Support-index construction helpers for TRex simulation demos.
# Mirrors R/trex_selector_methods/support_generators.R.
#
# Contents:
#   make_support_capped_spread()  — Deterministic evenly-spaced support.
#   make_support_random()         — Uniformly random support.
#
# ==============================================================================

import numpy as np


def make_support_capped_spread(s, p, max_gap=10):
    """
    Build an evenly-spaced support with a capped gap (deterministic).

    Mirrors R make_support_capped_spread():
        gap = min(p // s, max_gap)
        indices = {0, gap, 2*gap, ..., (s-1)*gap}  (0-based)

    Parameters
    ----------
    s       : int  Number of active predictors.
    p       : int  Total number of predictors.
    max_gap : int  Maximum gap between consecutive active indices. Default 10.

    Returns
    -------
    np.ndarray of int, shape (s,), 0-based, sorted ascending.

    Examples
    --------
    >>> make_support_capped_spread(10, 1000, 20)
    array([  0,  20,  40,  60,  80, 100, 120, 140, 160, 180])
    """
    gap = min(p // s, max_gap)
    return np.arange(s, dtype=np.intp) * gap


def make_support_random(s, p, seed):
    """
    Build a uniformly random support without replacement (stochastic).

    Mirrors R make_support_random():
        rng = np.random.default_rng(seed + 500_000)
        support = sorted sample of size s from {0, ..., p-1}

    The +500_000 offset keeps the support-draw RNG stream disjoint
    from the DGP RNG stream (which uses seed directly), and matches
    the R convention of set.seed(seed + 500000L).

    Parameters
    ----------
    s    : int  Number of active predictors.
    p    : int  Total number of predictors.
    seed : int  Base seed (500_000 will be added internally).

    Returns
    -------
    np.ndarray of int, shape (s,), 0-based, sorted ascending.
    """
    rng = np.random.default_rng(int(seed) + 500_000)
    return np.sort(rng.choice(p, size=s, replace=False).astype(np.intp))
